### PR TITLE
Fix MCX synthesis methods when the number of control qubits is 0

### LINKED ISSUE #15665: `synth_mcx_n_dirty_i15` with `num_ctrl_qubits=0` causes unbounded memory allocation and crash
### Environment

- **Qiskit version**: 2.3.0
- **Python version**: 3.11.14
- **Operating system**: Ubuntu 22.04

### What is happening?

Calling `synth_mcx_n_dirty_i15` with `num_ctrl_qubits=0` triggers unbounded memory allocation. The process continuously consumes memory and eventually crashes with a `core dumped` error.

### How can we reproduce the issue?

```python
from qiskit.synthesis.multi_controlled import synth_mcx_n_dirty_i15
qc = synth_mcx_n_dirty_i15(num_ctrl_qubits=0)
```
Running this code leads to rapidly increasing memory usage and eventually a crash.

### What should happen?

The function should either raise a clear `ValueError` for `num_ctrl_qubits=0` or handle this case explicitly, but must not allocate unbounded memory or crash.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

The PR extends MCX synthesis methods to correctly handle the edge case where ``num_ctrl_qubits`` is 0. 

Fixes #15665.

### Details and comments

While this is somewhat of an edge case, the synthesis functions (such as `synth_mcx_n_dirty_i15`) appear in our [documentation](https://quantum.cloud.ibm.com/docs/en/api/qiskit/synthesis). As no restrictions on the number of control qubits are stated, it makes sense to support all nonnegative values for ``num_ctrl_qubits`` for all documented synthesis methods.

In fact, some of the documented synthesis methods require at least 3 control qubits, with this PR a default implementation (consisting of an X-gate, CX-gate or CCX-gate respectively) is used for small values of ``num_ctrl_qubits`` (including 0).

In addition, an improved error message is raised when the number of control qubits passed to the synthesis methods is negative. 

Note that I did not change the HLS plugins: some of the plugins still return None when the number of control qubits is low. This is fine, thought, as the default plugin can handle all of the values.

### CODE REVIEW

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 1]
```diff
@@ -57,31 +57,71 @@ def synth_mcx_n_dirty_i15(
...
+    Raises:
+        QiskitError: if ``num_ctrl_qubits`` is illegal.
+
     References:
         1. Iten et. al., *Quantum Circuits for Isometries*, Phys. Rev. A 93, 032318 (2016),
            `arXiv:1501.06911 <https://arxiv.org/abs/1501.06911>`_
     """
+    if num_ctrl_qubits < 0:
+        raise QiskitError(
+            "The method synth_mcx_n_dirty_i15 cannot be called with a negative number of control qubits."
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: In fact, this method needs k-2 ancillas, so there is no reason to call it for k<3.

Comment (ShellyGarion) [reply qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps it's better to write in the docstring above (line 45):
For :math:`k\le 3` explicit efficient circuits are used instead that do not require ancillas.

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 2]
```diff
@@ -57,31 +57,71 @@ def synth_mcx_n_dirty_i15(
...
+        raise QiskitError("_synth_mcx_special_cases should be called with only 0, 1, or 2 cotrols.")
+
+
 def synth_mcx_n_clean_m15(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`k - 2`
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`k - 2`
     clean ancillary qubits with producing a circuit with :math:`2 * k - 1` qubits
     and at most :math:`6 * k - 6` CX gates, by Maslov [1].
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 3]
```diff
@@ -111,22 +151,33 @@ def synth_mcx_n_clean_m15(num_ctrl_qubits: int) -> QuantumCircuit:
 
 def synth_mcx_1_clean_b95(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using a single
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using a single
     clean ancillary qubit producing a circuit with :math:`k + 2` qubits and at most
     :math:`16 * k - 24` CX gates, by [1], [2].
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 4]
```diff
@@ -169,18 +220,30 @@ def synth_mcx_1_clean_b95(num_ctrl_qubits: int) -> QuantumCircuit:
 
 def synth_mcx_gray_code(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using the Gray code.
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using the Gray code.
 
     Produces a quantum circuit with :math:`k + 1` qubits. This method
     produces exponentially many CX gates and should be used only for small
     values of :math:`k`.
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 5]
```diff
@@ -331,8 +410,9 @@ def _linear_depth_ladder_ops(num_ladder_qubits: int) -> tuple[QuantumCircuit, li
 
 def synth_mcx_1_kg24(num_ctrl_qubits: int, clean: bool = True) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`1` ancillary qubit as
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`1` ancillary qubit as
     described in Sec. 5 of [1].
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 6]
```diff
@@ -380,9 +464,9 @@ def synth_mcx_1_kg24(num_ctrl_qubits: int, clean: bool = True) -> QuantumCircuit
 
 def synth_mcx_1_clean_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`1` clean ancillary qubit
-    producing a circuit with :math:`2k-3` Toffoli gates or :math:`6k-6` CX gates and depth
-    :math:`O(k)` as described in Sec. 5.1 of [1].
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`1` clean
+    ancillary qubit producing a circuit with :math:`2k-3` Toffoli gates or :math:`6k-6`
+    CX gates and depth :math:`O(k)` as described in Sec. 5.1 of [1].
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py:476]: Maybe add:    
For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively, and no ancillas are used.

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 7]
```diff
@@ -391,21 +475,30 @@ def synth_mcx_1_clean_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
...
 
 def synth_mcx_1_dirty_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`1` dirty ancillary qubit
-    producing a circuit with :math:`4k-8` Toffoli gates or :math:`12k-18` CX gates and depth
-    :math:`O(k)` as described in Sec. 5.3 of [1].
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`1` dirty
+    ancillary qubit producing a circuit with :math:`4k-8` Toffoli gates or :math:`12k-18`
+    CX gates and depth :math:`O(k)` as described in Sec. 5.3 of [1].
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 8]
```diff
@@ -488,8 +588,9 @@ def _build_logn_depth_ccx_ladder(
 
 def synth_mcx_2_kg24(num_ctrl_qubits: int, clean: bool = True) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`2` ancillary qubits.
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`2` ancillary qubits.
     as described in Sec. 5 of [1].
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 9]
```diff
@@ -553,9 +659,10 @@ def synth_mcx_2_kg24(num_ctrl_qubits: int, clean: bool = True) -> QuantumCircuit
 
 def synth_mcx_2_clean_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`2` clean ancillary qubits
-    producing a circuit with :math:`2k-3` Toffoli gates or :math:`6k-6` CX gates and
-    depth :math:`O(\log(k))` as described in Sec. 5.2 of [1].
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`2` clean
+    ancillary qubits producing a circuit with :math:`2k-3` Toffoli gates or :math:`6k-6`
+    CX gates and depth :math:`O(\log(k))` as described in Sec. 5.2 of [1].
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 10]
```diff
@@ -564,21 +671,30 @@ def synth_mcx_2_clean_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
...
 
 def synth_mcx_2_dirty_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`2` dirty ancillary qubits
-    producing a circuit with :math:`4k-8` Toffoli gates or :math:`12k-18` CX gates and depth
-    :math:`O(\log(k))` as described in Sec. 5.4 of [1].
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`2` dirty
+    ancillary qubits producing a circuit with :math:`4k-8` Toffoli gates or :math:`12k-18`CX
+    gates and depth :math:`O(\log(k))` as described in Sec. 5.4 of [1].
+    For :math:`k\le 2` the returned circuit contains a single X, CX or CCX gate respectively.
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: perhaps worth adding that no ancillas are used for k<3?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 11]
```diff
@@ -564,21 +671,30 @@ def synth_mcx_2_clean_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
...
     return synth_mcx_2_kg24(num_ctrl_qubits, clean=True)
 
 
 def synth_mcx_2_dirty_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
     r"""
-    Synthesize a multi-controlled X gate with :math:`k` controls using :math:`2` dirty ancillary qubits
-    producing a circuit with :math:`4k-8` Toffoli gates or :math:`12k-18` CX gates and depth
-    :math:`O(\log(k))` as described in Sec. 5.4 of [1].
+    Synthesize a multi-controlled X gate with :math:`k\ge 3` controls using :math:`2` dirty
+    ancillary qubits producing a circuit with :math:`4k-8` Toffoli gates or :math:`12k-18`CX
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]:
```suggestion
    ancillary qubits producing a circuit with :math:`4k-8` Toffoli gates or :math:`12k-18` CX
```

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 12]
```diff
@@ -213,11 +276,19 @@ def synth_mcx_noaux_v24(num_ctrl_qubits: int) -> QuantumCircuit:
...
+        QiskitError: if ``num_ctrl_qubits`` is illegal.
+
     References:
         1. Vale et. al., *Circuit Decomposition of Multicontrolled Special Unitary
            Single-Qubit Gates*, IEEE TCAD 43(3) (2024),
            `arXiv:2302.06377 <https://arxiv.org/abs/2302.06377>`_
     """
+    if num_ctrl_qubits < 0:
+        raise QiskitError(
+            "The method synth_mcx_noaux_v24 cannot be called with a negative number of control qubits."
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: you don't need to update the docstring of `synth_mcx_noaux_v24` ?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 13]
```diff
@@ -236,11 +307,19 @@ def synth_mcx_noaux_hp24(num_ctrl_qubits: int) -> QuantumCircuit:
...
+        QiskitError: if ``num_ctrl_qubits`` is illegal.
+
     References:
         1. Huang and Palsberg, *Compiling Conditional Quantum Gates without Using
            Helper Qubits*, PLDI (2024),
            <https://dl.acm.org/doi/10.1145/3656436>`_
     """
+    if num_ctrl_qubits < 0:
+        raise QiskitError(
+            "The method synth_mcx_noaux_hp24 cannot be called with a negative number of control qubits."
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: you don't need to update the docstring of `synth_mcx_noaux_hp24` ?

**Review by ShellyGarion — COMMENTED** [review]
@alexanderivrii - thanks for fixing it bug and making a consistent behaviour of all MCX synthesis functions. I have some minor comments.

Would you like to add a test that asserts that calling MCX with negative number of qubits raises an error? or is it too much?

**Comment by ShellyGarion** [comment]
Another comment: 
not this PR nor the following one handle MCMT gate.
in particular, `synth_mcmt_vchain` and `synth_mcmt_xgate`.
have you checked them?

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py (hunk 1 above)
Comment (alexanderivrii) [reply qiskit/synthesis/multi_controlled/mcx_synthesis.py]: I agree, there is no reason to call this function for $k<3$, however it is better to provide some a default implementation rather than throwing an error message.

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py (hunk 12 above)
Comment (alexanderivrii) [reply qiskit/synthesis/multi_controlled/mcx_synthesis.py]: I think we can leave the docstring as it, as the method does not use ancillas.

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py (hunk 13 above)
Comment (alexanderivrii) [reply qiskit/synthesis/multi_controlled/mcx_synthesis.py]: Same as above: I think we can leave the docstring as is, as the algorithm does not need ancillas.

**Comment by alexanderivrii** [comment]
Thanks Shelly for the review, I have addressed the documentation improvements in [b71d228](https://github.com/Qiskit/qiskit/pull/15672/commits/b71d228389eefeeecb49e807c540768f5454ad89).

> not this PR nor the following one handle MCMT gate.
in particular, synth_mcmt_vchain and synth_mcmt_xgate.
have you checked them?

I have a follow-up PR #15675 on handling MCMT gates, however I am not fully sure that extending MCMT to 0-controlled gates is needed. Also note that currently MCMT gates are documented as requiring at least one control qubit, and MCMT synthesis methods do give good error messages when called with an invalid number of controls.

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 14]
```diff
@@ -391,21 +481,31 @@ def synth_mcx_1_clean_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
...
+        QiskitError: if ``num_ctrl_qubits`` is illegal.
 
     References:
         1. Khattar and Gidney, Rise of conditionally clean ancillae for optimizing quantum circuits
         `arXiv:2407.17966 <https://arxiv.org/abs/2407.17966>`__
     """
 
+    if num_ctrl_qubits < 0:
+        raise QiskitError(
+            "The method synth_mcx_n_dirty_i15 cannot be called with a negative number of control qubits."
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: The method synth_mcx_1_clean_kg24

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 15]
```diff
@@ -414,12 +514,19 @@ def synth_mcx_1_dirty_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
...
-        QiskitError: If num_ctrl_qubits <= 2.
+        QiskitError: if ``num_ctrl_qubits`` is illegal.
 
     References:
         1. Khattar and Gidney, Rise of conditionally clean ancillae for optimizing quantum circuits
         `arXiv:2407.17966 <https://arxiv.org/abs/2407.17966>`__
     """
+    if num_ctrl_qubits < 0:
+        raise QiskitError(
+            "The method synth_mcx_n_dirty_i15 cannot be called with a negative number of control qubits."
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: The method synth_mcx_1_dirty_kg24

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 16]
```diff
@@ -499,15 +608,20 @@ def synth_mcx_2_kg24(num_ctrl_qubits: int, clean: bool = True) -> QuantumCircuit
...
+        QiskitError: if ``num_ctrl_qubits`` is illegal.
 
     References:
         1. Khattar and Gidney, Rise of conditionally clean ancillae for optimizing quantum circuits
         `arXiv:2407.17966 <https://arxiv.org/abs/2407.17966>`__
     """
 
+    if num_ctrl_qubits < 0:
+        raise QiskitError(
+            "The method synth_mcx_n_dirty_i15 cannot be called with a negative number of control qubits."
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: The method synth_mcx_2_kg24

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 17]
```diff
@@ -564,21 +680,31 @@ def synth_mcx_2_clean_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
...
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: The method synth_mcx_2_clean_kg24

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py [hunk 18]
```diff
@@ -587,12 +713,19 @@ def synth_mcx_2_dirty_kg24(num_ctrl_qubits: int) -> QuantumCircuit:
...
-        QiskitError: If num_ctrl_qubits <= 2.
+        QiskitError: if ``num_ctrl_qubits`` is illegal.
 
     References:
         1. Khattar and Gidney, Rise of conditionally clean ancillae for optimizing quantum circuits
         `arXiv:2407.17966 <https://arxiv.org/abs/2407.17966>`__
     """
+    if num_ctrl_qubits < 0:
+        raise QiskitError(
+            "The method synth_mcx_n_dirty_i15 cannot be called with a negative number of control qubits."
```
Comment (ShellyGarion) [inline qiskit/synthesis/multi_controlled/mcx_synthesis.py]: The method synth_mcx_2_dirty_kg24

**Review by ShellyGarion — COMMENTED** [review]
I found a few misprints of function names in error messages. 
Other than that, it's ready to go.

File: qiskit/synthesis/multi_controlled/mcx_synthesis.py (hunk 14 above)
Comment (alexanderivrii) [reply qiskit/synthesis/multi_controlled/mcx_synthesis.py]: oops copy pasting is evil

**Comment by alexanderivrii** [comment]
Many thanks Shelly! Fixed error messages in [3625776](https://github.com/Qiskit/qiskit/pull/15672/commits/36257764fcedd7526a0681de6731214c2c263764).

**Review by ShellyGarion — APPROVED** [review]
